core/internal/filetransfer/file_transfer_default.go

```diff
@@ -77,10 +78,23 @@ func (ft *DefaultFileTransfer) Upload(task *Task) error {
 		)
 	}
 
-	task.Size = stat.Size()
+	if task.Offset > 0 || (task.Length > 0 && task.Length < stat.Size()) {
+		// Treat this as a range upload if we start somewhere other than the beginning or
+		// if a length is specified and it's less than the file size.
+		if task.Offset+task.Length > stat.Size() {
+			// If the range exceeds the file size, there was some kind of error upstream.
+			return fmt.Errorf("file transfer: upload: offset + length exceeds the file size")
+		}
+		sectionReader := io.NewSectionReader(file, task.Offset, task.Length)
+		reader = sectionReader
+		task.Size = task.Length
+	} else {
+		reader = file
+		task.Size = stat.Size()
+	}
```

Comment:
It seems it would be simpler to always use a `SectionReader`. When `task.Length == 0`, we can use `stat.Size() - task.Offset` (and on that topic, it's a bit confusing that there's both `Length` and `Size`, and that `Length` is an input but `Size` is inferred...)

Reply:
Good point. I reworked it to eliminate Length and reuse Size for this purpose. If you think that's overloading `Size` too much I will revert to use Length as well.

On a related note, there are certain corner cases we can't handle either way; in this case, it's impossible to upload a segment of length 0 starting at offset 0 from a file with >0 bytes; trying to do so will upload the entire file. However, I can't think of a reason someone would want to do this.